lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 9.2
+ */
+public class RequestExpirationScheduler {
+
+   private static final Log log = LogFactory.getLog(RequestExpirationScheduler.class, Log.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private final ScheduledExecutorService scheduledExecutorService;
+   private final ConcurrentHashMap<String, ScheduledRequest> scheduledRequests = new ConcurrentHashMap<>();
+
+   public RequestExpirationScheduler(ScheduledExecutorService scheduledExecutorService) {
+      this.scheduledExecutorService = scheduledExecutorService;
+   }
+
+   class ScheduledRequest {
+      private String requestId;
+      private CompletableFuture<Boolean> request;
+      private ScheduledFuture<?> scheduledFuture;
+
+      public ScheduledRequest(String requestId, CompletableFuture<Boolean> request, ScheduledFuture<?> scheduledFuture) {
+         this.requestId = requestId;
+         this.request = request;
+         this.scheduledFuture = scheduledFuture;
+      }
+
+      public String getRequestId() {
+         return requestId;
+      }
+
+      public CompletableFuture<Boolean> getRequest() {
+         return request;
+      }
+
+      public ScheduledFuture<?> getScheduledFuture() {
+         return scheduledFuture;
+      }
+   }
+
+   /**
+    * Schedules a request for completion
+    *
+    * @param requestId, the unique identifier if the request
+    * @param request, the request
+    * @param time, time expressed in long
+    * @param unit, {@link TimeUnit}
+    */
+   public void scheduleForCompletion(String requestId, CompletableFuture<Boolean> request, long time, TimeUnit unit) {
+      if (request.isDone()) {
+         if(trace) {
+            log.tracef("Request[%s] is not scheduled because is already done", requestId);
+         }
+         return;
+      }
+
+      if (scheduledRequests.containsKey(requestId)) {
+         if(trace) {
+            log.tracef("Request[%s] is not scheduled because it is already scheduled", requestId);
+         }
+         return;
+      }
+
+      if(trace) {
+         log.tracef("Request[%s] being scheduled to be completed in [%d, %s]", requestId, time, unit);
+      }
+
+      ScheduledFuture<?> scheduledFuture = scheduledExecutorService.schedule(() -> {
+         request.complete(false);
+         scheduledRequests.remove(requestId);
+      }, time, unit);
+
+      scheduledRequests.putIfAbsent(requestId, new ScheduledRequest(requestId, request, scheduledFuture));
+   }
+
+   /**
+    * Aborts the scheduled request
+    *
+    * @param requestId, unique identifier of the request
+    */
+   public void abortScheduling(String requestId) {
+      if(trace) {
+         log.tracef("Request[%s] abort scheduling", requestId);
+      }
+      ScheduledRequest scheduledRequest = scheduledRequests.get(requestId);
+      if (scheduledRequest != null && scheduledRequest.request.isDone()) {
```

Comment:
We shouldn't have to care if the request is done or not, just cancel the future.